app/main.py

```diff
@@ -1,34 +1,90 @@
+            end: tuple,
+            is_drowned: bool = False
+    ) -> None:
+        self.is_drowned = is_drowned
+        length = abs(start[0] - end[0]) + abs(start[1] - end[1]) + 1
+        self.decks = []
+        for i in range(length):
+            if start[0] == end[0]:
+                deck = Deck(start[0], start[1] + i)
+            else:
+                deck = Deck(start[0] + i, start[1])
+            self.decks.append(deck)
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                deck.is_alive = False
+        if all([not deck.is_alive for deck in self.decks]):
+            self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+            self,
+            ships: list[(tuple[int, int], tuple[int, int])]
+    ) -> None:
+        self.ships = [Ship(start=ship[0], end=ship[1]) for ship in ships]
+        self.field = {}
+        for ship in self.ships:
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def fire(self, location: tuple) -> str:
+        if location in self.field:
+            ship = self.field[location]
+            ship.fire(location[0], location[1])
+            if ship.is_drowned:
+                return "Sunk!"
+            return "Hit!"
+        else:
+            return "Miss!"
+
+    def print_field(self) -> None:
+        for row in range(10):
+            for column in range(10):
+                if (row, column) in self.field:
+                    ship = self.field[(row, column)]
+                    if ship.is_drowned:
+                        print("x", end="  ")
+                    else:
+                        if ship.get_deck(row, column).is_alive:
+                            print(u"\u25A1", end="  ")
+                        else:
+                            print("*", end="  ")
+                else:
+                    print("~", end="  ")
+            print()
```

Comment:
redundant print

Reply:
this print jumps to a new row, so the result is 10x10 grid instead of 1x100 row